### PR TITLE
Add PParams for Conway

### LINKED ISSUE #3146: new Conway protocol parameters
new Conway protocol parameters:

- all the governance voting thresholds
- constitutional committee term limits
- governance action expiration
- governance action deposit (govDeposit)
- DRep deposit amount (drepDeposit)
- DRep activity period (drepActivity)
- minimal constitutional committee size
- maximum term limit (in epochs) of the constitutional committee

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Add `PParams` for `Conway`

Fixes #3146 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/cardano-ledger-conway.cabal
Comment (lehins) [inline eras/conway/impl/cardano-ledger-conway.cabal]: Patch version makes sense only for non-visible changes. This PR changes a type fimily instance from BabbagePParams to a new ConwayPParams, which is a breaking change

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]:
```suggestion

  -- New ones for Conway
  , cppVotingThresholds :: !(HKD f (Map.Map Natural Natural))
  -- ^ Thresholds for SPOs and DReps votes
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Protocol version is not counted because it is unrolled using a group decoding
```suggestion
    encodeListLen (28 + listLen cppProtocolVersion)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]:
```suggestion
    decodeRecordNamed "PParams" (\pp -> 28 + fromIntegral (listLen (cppProtocolVersion pp))) $ do
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs]: Let's create concrete two record types for this instead of a single pparam with a  `Map`, this will improve safety of our rules.

This would look something like this:

```haskell
data PoolVotingThresholds = PoolVotingThresholds
  { pvtMotionNoConfidence :: !UnitInterval
  , ...
  , pvtHardForkInitiation :: !UnitInterval
  }
  deriving (Show, Generic)

data DRepVotingThresholds = DRepVotingThresholds
  { dvtMotionNoConfidence :: !UnitInterval
  , ...
  , dvtUpdateToConstitution :: !UnitInterval
  , dvtHardForkInitiation :: !UnitInterval
  , ...
  }
  deriving (Show, Generic)

data CownayPParams era = ConwayPParams
  { ...
  , ccpPoolVotingThresholds :: !PoolVotingThresholds
  , ccpDRepVotingThresholds :: !DRepVotingThresholds
  , ...
  }
  deriving (Show, Generic)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: Why is this change necessary?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: Same here?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: And here?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: Why did you need to move it into PParams module?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs]: There are no other committees, except constitutional committees in ledger, so lets avoid the CC abbreviation and just use committee instead.
```suggestion
  hkdMinCommitteSizeL :: HKDFunctor f => Lens' (PParamsHKD f era) (HKD f Natural)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs]:
```suggestion
  hkdCommitteeTermLimitL :: HKDFunctor f => Lens' (PParamsHKD f era) (HKD f Natural)
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Naturally, my comment about CC applies to all of the field names too

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Seems like we should either mention all the new ones here or none? it's a little confusing to only mention one , no?

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: s/ `loosing`/`losing`

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Is it better for this to reside in `PParams` instead of `Governance` module?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs]:
```suggestion
  hkdGovActionExpirationL :: HKDFunctor f => Lens' (PParamsHKD f era) (HKD f Natural)
  hkdGovActionDepositL :: HKDFunctor f => Lens' (PParamsHKD f era) (HKD f Coin)
```

**Review by lehins — CHANGES_REQUESTED** [review]
This looks great. Voting thresholds need some work still. There is a related spec PR was opened earlier today: https://github.com/input-output-hk/formal-ledger-specifications/pull/134

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Thanks for the catch!

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: This was moved due to a cyclical import problem that I started discussing on Slack. This PR is on hold now, until the module for `Conway.Governance` is split properly, in a separate PR.

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Oh, thanks!

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs]: Got it! :raised_hands:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs:578]: Noting here: This was moved due to a cyclical import problem that I started discussing on Slack with @lehins . This PR is on hold now, until the module for `Conway.Governance` is split properly, in a separate PR.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: Noting here: This was moved due to a cyclical import problem that I started discussing on Slack with @lehins . This PR is on hold now, until the module for `Conway.Governance` is split properly, in a separate PR.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs]: Got it! :raised_hands:

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (Soupstraw) [reply eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: `Coders` saves the day :wink:

Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Except you can't use Group with coders. But because all previous eras unroll ProtVer, we also have to do it for Conway

**Comment by Soupstraw** [comment]
LGTM, good job @aniketd!

**Review by Soupstraw — APPROVED** [review]

File: eras/babbage/impl/CHANGELOG.md
Comment (lehins) [inline eras/babbage/impl/CHANGELOG.md:3]: We need to add a lower bound in cardano-ledger-conway on this version, since it now needs this function.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Core.hs:26]: We are missing the same set of lenses for PParamsUpdates, i.e. with `ppu*` prefix

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: Looks like we need some roundtrip tests for pparmas and pparamsupdates
```suggestion
    !> omitStrictMaybe 25 (cppPoolVotingThresholds ppup) encCBOR
    !> omitStrictMaybe 26 (cppDRepVotingThresholds ppup) encCBOR
    !> omitStrictMaybe 27 (cppMinCommitteeSize ppup) encCBOR
    !> omitStrictMaybe 28 (cppCommitteeTermLimit ppup) encCBOR
    !> omitStrictMaybe 29 (cppGovActionExpiration ppup) encCBOR
    !> omitStrictMaybe 30 (cppGovActionDeposit ppup) encCBOR
    !> omitStrictMaybe 31 (cppDRepDeposit ppup) encCBOR
    !> omitStrictMaybe 32 (cppDRepActivity ppup) encCBOR
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: redundant
```suggestion
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs]: You reintroduced back the bug that I tried fixing so hard in
 #3440 and 
https://github.com/input-output-hk/cardano-ledger/pull/3297/files#diff-26cafa197bd48b2a0468db9e9da85fd570c45100657e8c87e5b3cad2d2acbd6d

**Review by lehins — CHANGES_REQUESTED** [review]
Could you please add:
* roundtrip tests for PParams and PParamsUpdates
* CDDL tests and spec for PParamsUpdates

There was a few bugs that I caught with regards to serialization, so tests would help us catch them and possibly others

**Review by teodanciu — APPROVED** [review]
Looks good to me, just wondering about the `minimumAVS` param that appears in the spec but not in the implementation, but probably something that I missed.

File: eras/conway/impl/cardano-ledger-conway.cabal
Comment (teodanciu) [inline eras/conway/impl/cardano-ledger-conway.cabal:72]: I think we need to increase the version of cardano-ledger-conway package? In case it's not already increased after rebasing

Comment (teodanciu) [reply eras/conway/impl/cardano-ledger-conway.cabal:72]: Looks like master has 1.6.1.1 right now, but depends on when this gets merged.

**Review by teodanciu — COMMENTED** [review]
~Before merging, check that the version of cardano-ledger-conway is more than 1.6.1.1  (either by changing it in this PR or by rebasing, whichever is needed at the time of merging)~ 
Actually nevermind,  1.6.1.1 is not on chap, my mistake  (unless... someone bumps to 1.6.1.1 and releases to chap before you merge this :laughing: )

**Review by lehins — APPROVED** [review]
Awesome stuff

**Pull request merged**
